src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTarget.java

```diff
@@ -5093,8 +5095,38 @@ protected void onChunkWritten(String chunk,
          else if (content.trim().startsWith("@"))
             wordWrap.setWrappingEnabled(true);
 
+         if (content.matches("^\\s*```")) 
+         {
+            outsideMarkdown = !outsideMarkdown;
+            wordWrap.setWrappingEnabled(outsideMarkdown);
+         }
+
+         boolean isWrappingEnabled = wordWrap.getWrappingEnabled();
+         boolean bullet = content.matches("^\\s*[-*].*");
+         if (bullet)
+            wordWrap.setWrappingEnabled(false);
+
+         boolean macroStart = content.matches("^\\s*\\\\[a-zA-Z]+\\{.*");
+         if (macroStart) 
+         {
+            macroDepth = macroDepth + 1;
+            wordWrap.setWrappingEnabled(false);
+         }
+
+         boolean macroEnd = content.matches("^\\s*\\}.*");
+         if (macroEnd) 
+         {
+            macroDepth = macroDepth - 1;
```

Comment:
We should guard against unmatched `}` here (make sure `macroDepth` is always >= 0)